### PR TITLE
guest plugins

### PR DESCRIPTION
**Pull request by lacraig2**
This is a _draft_ PR for tracking the new feature: guest plugins.

The PR includes several new PANDA plugins:
- `linjector` - A hypervisor-based executable injector.
- `guest_plugin_manager` - A general interface for PANDA plugins to interact with guest plugins.
- `hyperfuse` - An example guest plugin for PANDA which exposes the internal guest file system to the host file system via FUSE.
- `guest_shell` - An example guest plugin for PANDA which exposes an injected command line interface for the guest to the host.

TODO:
- [x] Build generic interface for `panda_require` from rust and make use of it for `guest_plugin_manager` to load `linjector`
- [ ] linjector port to MIPS and aarch64
- [ ] Fix linjector on currently "supported" but broken architectures
  - [x] ARM
  - [ ] i386 
- [x] Add a PPP callback for "linjector injected" and a consumer in `guest_plugin_manager`
- [x] Add a PPP callback for "guest agent loaded" and add consumer in PANDA guest plugins
- [x] Remove hardcoded paths from linjector/hyperfuse
- [x] Add a guest plugin retrieval-by-name mechanism
  - [x] Add load-by-name to guest_plugin_manager
  - [x] Add bindings to panda-rs  
- [x] Add guest plugin building to the build system (prereq for last 2)
  - [x] Add out-of-tree guest plugin building
- [x] Add a Rust attribute for wrapping the message_recv callback
- [x] PANDA API documentation
  - [x] Rust documentation ([link](https://docs.rs/panda-re/latest/panda/plugins/guest_plugin_manager/index.html))
  - [x] Add host plugin example ([link](https://github.com/panda-re/panda/blob/linjector_base4/panda/plugins/guest_plugin_example/src/lib.rs), [permalink](https://github.com/panda-re/panda/blob/44fc6a212572e24eaba94727b3d07cf9db89e3bc/panda/plugins/guest_plugin_example/src/lib.rs))
  - [x] Add Guest Plugin Tutorial/Guide
  - [x] Add README to linjector
- [ ] guest plugin API guide
  - [x] Add guest plugin example ([link](https://github.com/panda-re/panda/blob/linjector_base4/panda/guest_plugins/rust_example/src/main.rs), [permalink](https://github.com/panda-re/panda/blob/44fc6a212572e24eaba94727b3d07cf9db89e3bc/panda/guest_plugins/rust_example/src/main.rs))
- [x] move `guest_plugin_path` function into panda-rs
- [x] remove hardcoded path from `rust_example` guest plugin
  - [x] setup hyperchannels library for commit-agnostic versioning
    - [x] move to its own repo
    - [x] publish to crates.io
- [x] Rewrite PANDA-side guest_shell plugin for readability
- [ ] hyperfuse improvements
  - [x] Implement symlink resolution
    - [x] Rewrite list caching 
      - [x] Rework to not invalidate on hit (links are resolved twice by ls, for example)
      - [x] Pre-fetch lookups of symlinks on directory listing
  - [x] Fix I/O error for `/proc` and `/sys`
  - [ ] Investigate which unimplemented features are important
  - [x] Have mount point be an environment variable

### CODE REVIEW

**Comment by jamcleod** [comment]
Instructions for trying hyperfuse/guest_shell today:

### Setup

1. Clone both `panda-re/panda` and `panda-re/igloo-internal`
2. Checkout the `linjector_base4` branch of panda
3. Checkout the `guest_agent` branch of igloo-internal
4. Set the `EXTRA_GUEST_PLUGINS_PATH` environment variable to point to `igloo-internal/pie_idea/guest_code`
5. Ensure you have the musl target installed for Rust:

```
rustup target add x86_64-unknown-linux-musl
```

6. Build PANDA for x86_64 (other architectures are currently broken)

```
cd build
../build.sh x86_64-softmmu
```

### Using hyperfuse

1. Create a directory to use as your mount point for the guest filesystem
2. Set the `HYPERFUSE_MOUNT` environment variable to point to that directory
3. Run `panda/python/examples/hyperfuse.py`

Give it a second to start up, then your mount point should now mirror the guest filesystem's state.

### Using guest_shell

1. Run `panda/python/examples/guest_shell.py`
2. In another terminal (the one you want to use as the shell) run `panda/plugins/guest_shell/guest_shell_pty.sh`
3. Give it a second to start up. You should see a `#` prompt in your second terminal and you can begin using the shell

**Comment by jamcleod** [comment]
```python 
@panda.ppp("linjector", "before_guest_inject")
def linjector_finish(cpu):
    print("!!!! Injection finished !!!!")

@panda.ppp("guest_plugin_manager", "on_guest_agent_load")
def agent_loaded(cpu):
    print("!!!! Guest agent loaded !!!!")
```
Examples of the added PPP callbacks

File: panda/docs/guest_plugins.md [hunk 1]
```diff
@@ -0,0 +1,306 @@
...
+#### Building
+
+To build, we go through the standard PANDA build process:
+
+1. Create a `build` folder in the root of your clone of PANDA if you haven't already
+2. From within the build folder run `../build.sh`. In our case, we are only going to be testing on x86_64 while developing, so we can save ourselves some build time by only building a single architecture: `../build.sh x86_64-softmmu`
+
+#### Testing Our Guest Plugin
+
+The easiest way to test our plugin will be with a pypanda script. An example script can be found in [`panda/plugins/guest_plugin_example/try_it.py`](/panda/panda/plugins/guest_plugin_example/try_it.py).
```
Comment (lacraig2) [inline panda/docs/guest_plugins.md]: link broken

File: panda/dependencies/ubuntu:18.04_build.txt [hunk 2]
```diff
@@ -25,6 +25,9 @@ protobuf-compiler
 python3-dev
 zip
 
+# hyperfuse build deps
+libfuse-dev
```
Comment (lacraig2) [inline panda/dependencies/ubuntu:18.04_build.txt:29]: Does this need to be in the other ubuntu dependency files?

File: panda/guest_plugins/guest_daemon/src/main.rs [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    LoadPlugin = 0,
+}
+
+struct Packet {
+    kind: PacketKind,
+    payload: Vec<u8>,
+}
+
+fn read_packet(reader: &mut RawChannel) -> Option<Packet> {
+    //let plugin_file = MemfdOptions::new()
```
Comment (lacraig2) [inline panda/guest_plugins/guest_daemon/src/main.rs]: Do we need these large sections of commented code?

File: panda/plugins/guest_plugin_manager/src/interface/daemon_manager.rs [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        if let Ok(binary) = std::fs::read(binary_path) {
+            for chunk in binary.chunks(4096){
+                // let mut buf =
+                    // u32::to_le_bytes(PacketKind::LoadPlugin as u32).to_vec();
+                // buf.extend(u32::to_le_bytes(chunk.len() as u32).to_vec());
+                let cd = CHANNEL_DESC.get().unwrap();
+                // publish_message_to_guest(*cd, buf);
+                publish_message_to_guest(*cd, chunk.to_owned());
+            }
+            // let mut buf =
```
Comment (lacraig2) [inline panda/plugins/guest_plugin_manager/src/interface/daemon_manager.rs]: Another here with lots of commented code.

**Review by lacraig2 — COMMENTED** [review]
Looks good so far. Some small things to fix.

I'll test functionality in another review after these fixes get merged.

**Pull request merged**